planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite
+    count(*) -> count(not null column)
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(constant)
+
+Pattern:
+LogcialAggregation: count(constant)
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns from datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the narrowest not null column from datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns from datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the narrowest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
```

Comment:
Any benchmark for this case? In TiFlash, there are extra three columns <row_id, version, delmark> which are used for MVCC. Therefore, if there are multi version of data, we need to scan <row_id, version, delmark> each query additionally. For `count(1)` we need to scan <row_id, version, delmark> and return row_id to aggregation, but for `count(k)` we need to scan <row_id, version, delmark, k1> and return k1 to aggregation. So the performance may degrade when there are multi version of data in TiFlash and k1 is wide.

Reply:
I am doing performance testing.
Do you mean that in the case of multiple versions, even if the count(*) does not select the row_id column, the final read will still read the row_id column?

Reply:
yes, we need to use it to filter out other data of other versions.